GillianCore/engine/FOLogic/Reduction.ml

```diff
@@ -2412,638 +2636,84 @@ and substitute_for_list_length (pfs : PFS.t) (le : Expr.t) : Expr.t =
-            let vb0 = Z.sub n vb1 in
-            Formula.And
-              (Eq (LVar b1, Lit (Int vb1)), Eq (LVar b0, Lit (Int vb0)))
-      | Eq
-          ( BinOp (BinOp (Lit (Int tfs), ITimes, LVar b1), IPlus, LVar b0),
-            Lit (Int n) )
-        when top_level && Z.equal tfs _256
-             && PFS.mem pfs (ILessEq (Expr.zero_i, LVar b0))
-             && PFS.mem pfs (ILessEq (Expr.zero_i, LVar b1))
-             && PFS.mem pfs (ILess (LVar b0, Lit (Int _256)))
-             && PFS.mem pfs (ILess (LVar b1, Lit (Int _256))) ->
-          if Z.gt n _65535 then False
-          else
-            let vb1 = Z.div n _256 in
-            let vb0 = Z.sub n vb1 in
-            Formula.And
-              (Eq (LVar b1, Lit (Int vb1)), Eq (LVar b0, Lit (Int vb0)))
-      | Eq (BinOp (e, FTimes, Lit (Num x)), Lit (Num 0.)) when x <> 0. ->
-          Eq (e, Lit (Num 0.))
-      | Eq (BinOp (e, ITimes, Lit (Int x)), Lit (Int n))
-        when Z.equal n Z.zero && not (Z.equal x Z.zero) -> Eq (e, Expr.zero_i)
-      | Eq (BinOp (Lit (Num x), FTimes, e), Lit (Num 0.)) when x <> 0. ->
-          Eq (e, Lit (Num 0.))
-      | Eq (BinOp (Lit (Int x), ITimes, e), Lit (Int z))
-        when Z.equal z Z.zero && not (Z.equal x Z.zero) -> Eq (e, Expr.zero_i)
-      | Eq (Lit (LList ll), Lit (LList lr)) -> if ll = lr then True else False
-      | Eq (EList le, Lit (LList ll)) | Eq (Lit (LList ll), EList le) ->
-          if List.length ll <> List.length le then False
-          else if ll = [] then True
-          else
-            let eqs = List.map2 (fun x y -> Formula.Eq (x, Lit y)) le ll in
-            let conj =
-              List.fold_left
-                (fun ac x -> Formula.And (ac, x))
-                (List.hd eqs) (List.tl eqs)
-            in
-            conj
-      | Eq (EList ll, EList lr) ->
-          if List.length ll <> List.length lr then False
-          else if ll = [] then True
-          else
-            let eqs = List.map2 (fun x y -> Formula.Eq (x, y)) ll lr in
-            let conj =
-              List.fold_left
-                (fun ac x -> Formula.And (ac, x))
-                (List.hd eqs) (List.tl eqs)
-            in
-            conj
-      | Eq (left_list, right_list)
-        when (match
-                ( Typing.type_lexpr gamma left_list,
-                  Typing.type_lexpr gamma right_list )
-              with
-             | (Some Type.ListType, _), (Some Type.ListType, _) -> true
-             | _ -> false)
-             &&
-             match
-               fe
-                 (Expr.Infix.( - )
-                    (Expr.list_length left_list)
-                    (Expr.list_length right_list))
-             with
-             | Expr.Lit (Int k) when not (Z.equal k Z.zero) -> true
-             | _ -> false ->
-          (* If we have two lists but can reduce the equality of their lengths to false,
-             then we know the lists cannot be equal*)
-          False
```

Comment:
Removed this, which seems to slightly improve perf without causing any observable difference.